core/src/main/java/org/stellar/anchor/sep31/Sep31Service.java

```diff
@@ -77,42 +75,65 @@ public Sep31InfoResponse getInfo() {
   public Sep31PostTransactionResponse postTransaction(
       JwtToken jwtToken, Sep31PostTransactionRequest request) throws AnchorException {
     Context.reset();
-
     Context.get().setRequest(request);
     Context.get().setJwtToken(jwtToken);
 
-    validateAsset();
-
     AssetInfo assetInfo = assetService.getAsset(request.getAssetCode(), request.getAssetIssuer());
+    if (assetInfo == null) {
+      // the asset is not supported.
+      throw new BadRequestException(
+          String.format(
+              "asset %s:%s is not supported.", request.getAssetCode(), request.getAssetIssuer()));
+    }
+    Context.get().setAsset(assetInfo);
 
     // Pre-validation
     validateAmount(request.getAmount());
     validateLanguage(appConfig, request.getLang());
+    if (request.getFields() == null) {
+      throw new BadRequestException("'fields' field cannot be empty");
+    }
     validateRequiredFields(assetInfo.getCode(), request.getFields().getTransaction());
     validateSenderAndReceiver();
-    validateKyc();
     preValidateQuote();
 
     // Query the fee
-    queryFee();
+    updateFee();
 
     AssetInfo asset = Context.get().getAsset();
+    Amount fee = Context.get().getFee();
     Sep31Transaction txn =
         new Sep31TransactionBuilder(sep31TransactionStore)
             .id(generateSepTransactionId())
             .status(SepTransactionStatus.PENDING_SENDER.toString())
-            .stellarAccountId(asset.getDistributionAccount())
-            .amountInAsset(assetInfo.getAssetName())
-            .amountIn(request.getAmount())
+            .statusEta(null)
+            .amountFee(fee.getAmount())
+            .amountFeeAsset(fee.getAsset())
+            .startedAt(Instant.now())
+            .completedAt(null)
```

Comment:
Can we set the builder default values to `null` to reduce the verbosity.

Reply:
My intention here was to be verbose, actually. There were some fields that should be set but were not. Since there are so many fields, it's easy to miss something so I wanted to make it explicit which fields we're setting and which ones we're purposely not setting (the null ones).

This approach helps me answer the following question just by glancing at the code: "are we addressing all the fields?".

Reply:
I'm open to discussing it further, just let me know.